
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Pipeline cluster &#8212; cellhub 0.1 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/alabaster.css" />
    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Pipeline dehash" href="pipeline_dehash.html" />
    <link rel="prev" title="Pipeline Cell QC" href="pipeline_cell_qc.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <span class="target" id="module-cellhub.pipeline_cluster"></span><section id="pipeline-cluster">
<h1>Pipeline cluster<a class="headerlink" href="#pipeline-cluster" title="Permalink to this heading">¶</a></h1>
<section id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this heading">¶</a></h2>
<p>This pipeline performs clustering of integrated single cell datasets. Starting from an anndata
object with integrated coordinates (e.g. from Harmony or scVI) the pipeline:</p>
<ul class="simple">
<li><p>Computes the neighbour graph using the HNSW alogrithm</p></li>
<li><p>Performs UMAP compuation and Leiden clustering (with ScanPy)</p></li>
<li><p>Visualises QC statistics on the UMAPs and by cluster</p></li>
<li><p>Visualises singleR results</p></li>
<li><p>Finds cluster marker genes (using the ScanPy ‘rank_genes_groups’ function)</p></li>
<li><p>Performs pathway analysis of the cluster phenotypes (with gsfisher)</p></li>
<li><p>Prepares marker gene and summary reports</p></li>
<li><p>Export an anndata for viewing with cellxgene</p></li>
</ul>
<p>For plotting of data in R, the pipeline saves the input anndata in loom format and
reads data in R with the loomR library.</p>
</section>
<section id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this heading">¶</a></h2>
<p>See <a class="reference internal" href="../Installation.html"><span class="doc">Installation</span></a> and <a class="reference internal" href="../Usage.html"><span class="doc">Usage</span></a> on general
information how to use CGAT pipelines.</p>
<section id="configuration">
<h3>Configuration<a class="headerlink" href="#configuration" title="Permalink to this heading">¶</a></h3>
<p>It is recommended to perform the clustering in a new directory.</p>
<p>The pipeline requires a configured <code class="file docutils literal notranslate"><span class="pre">pipeline_cluster.yml</span></code> file.</p>
<p>Default configuration files can be generated by executing:</p>
<blockquote>
<div><p>python &lt;srcdir&gt;/pipeline_cluster.py config</p>
</div></blockquote>
</section>
<section id="inputs">
<h3>Inputs<a class="headerlink" href="#inputs" title="Permalink to this heading">¶</a></h3>
<p>The pipeline starts from an anndata object with the following structure.</p>
<ul class="simple">
<li><p>anndata.var.index -&gt; gene names - these will be displayed in the plots</p></li>
<li><p>anndata.var.gene_ids -&gt; ensembl gene ids, optional but required for pathway analysis</p></li>
<li><p>anndata.X -&gt; scaled data (a dense matrix)</p></li>
<li><p>anndata.layers[“counts”] -&gt; raw counts (a sparse matrix)</p></li>
<li><p>anndata.layers[“log1p”] -&gt; total count normalised, 1og1p transformed data (a sparse matrix)</p></li>
<li><p>anndata.obs -&gt; metadata (typically passed through from original cellhub object)</p></li>
<li><p>anndata.obsm[“X_rdim_name”] -&gt; containing the integrated coordinates (where “rdim_name” matches
the “runspec_rdim_name” parameter). TODO: rename this parameter.</p></li>
</ul>
<p>It is strongly recommended to retain the information for all of the genes in all of the matrices
(i.e. do not subset to HVGs!). This is important for marker gene discovery and pathway analysis.</p>
</section>
</section>
<section id="pipeline-output">
<h2>Pipeline output<a class="headerlink" href="#pipeline-output" title="Permalink to this heading">¶</a></h2>
<p>The pipeline produces the following outputs:</p>
<ol class="arabic simple">
<li><p>Summary report</p></li>
</ol>
<ul class="simple">
<li><p>Containing the overview UMAP plots, visualisation of QC and SingleR information</p></li>
<li><p>The result of the per-cluster pathway analysis</p></li>
</ul>
<ol class="arabic simple" start="2">
<li><p>Marker report</p></li>
</ol>
<ul class="simple">
<li><p>Containing heatmaps, violin plots, expression dotplots, MA and volcano plots for
each cluster.</p></li>
</ul>
<ol class="arabic simple" start="3">
<li><p>xlsx spreadsheets for the marker gene and pathway results</p></li>
<li><p>Anndata objects ready to be viewed with cellxgene</p></li>
</ol>
</section>
<section id="code">
<h2>Code<a class="headerlink" href="#code" title="Permalink to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_cluster.taskSummary">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_cluster.</span></span><span class="sig-name descname"><span class="pre">taskSummary</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_cluster.taskSummary" title="Permalink to this definition">¶</a></dt>
<dd><p>Make a summary of optional tasks that will be run</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_cluster.preflight">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_cluster.</span></span><span class="sig-name descname"><span class="pre">preflight</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_cluster.preflight" title="Permalink to this definition">¶</a></dt>
<dd><p>Preflight sanity checks.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_cluster.metadata">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_cluster.</span></span><span class="sig-name descname"><span class="pre">metadata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_cluster.metadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Export the metadata (obs) from the source anndata
for use in the plotting tasks</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_cluster.loom">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_cluster.</span></span><span class="sig-name descname"><span class="pre">loom</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_cluster.loom" title="Permalink to this definition">¶</a></dt>
<dd><p>Export the data to the loom file format. This is used
as an exchange format for plotting in R.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_cluster.neighbourGraph">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_cluster.</span></span><span class="sig-name descname"><span class="pre">neighbourGraph</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_cluster.neighbourGraph" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the neighbor graph.
A miniminal anndata is saved for UMAP computation and clustering.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_cluster.scanpyCluster">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_cluster.</span></span><span class="sig-name descname"><span class="pre">scanpyCluster</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_cluster.scanpyCluster" title="Permalink to this definition">¶</a></dt>
<dd><p>Discover clusters using ScanPy.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_cluster.cluster">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_cluster.</span></span><span class="sig-name descname"><span class="pre">cluster</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_cluster.cluster" title="Permalink to this definition">¶</a></dt>
<dd><p>Post-process the clustering result.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_cluster.compareClusters">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_cluster.</span></span><span class="sig-name descname"><span class="pre">compareClusters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_cluster.compareClusters" title="Permalink to this definition">¶</a></dt>
<dd><p>Draw a dendrogram showing the relationship between the clusters.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_cluster.clustree">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_cluster.</span></span><span class="sig-name descname"><span class="pre">clustree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_cluster.clustree" title="Permalink to this definition">¶</a></dt>
<dd><p>Run clustree.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_cluster.paga">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_cluster.</span></span><span class="sig-name descname"><span class="pre">paga</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_cluster.paga" title="Permalink to this definition">¶</a></dt>
<dd><p>Run partition-based graph abstraction (PAGA)
see: <a class="reference external" href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1663-x">https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1663-x</a></p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_cluster.UMAP">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_cluster.</span></span><span class="sig-name descname"><span class="pre">UMAP</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_cluster.UMAP" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the UMAP layout.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_cluster.plotRdimsFactors">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_cluster.</span></span><span class="sig-name descname"><span class="pre">plotRdimsFactors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infiles</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_cluster.plotRdimsFactors" title="Permalink to this definition">¶</a></dt>
<dd><p>Visualise factors of interest on the UMAP.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_cluster.plotRdimsClusters">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_cluster.</span></span><span class="sig-name descname"><span class="pre">plotRdimsClusters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_cluster.plotRdimsClusters" title="Permalink to this definition">¶</a></dt>
<dd><p>Visualise the clusters on the UMAP</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_cluster.plotRdimsSingleR">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_cluster.</span></span><span class="sig-name descname"><span class="pre">plotRdimsSingleR</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_cluster.plotRdimsSingleR" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot the SingleR primary identity assignments on a UMAP</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_cluster.plotSingleR">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_cluster.</span></span><span class="sig-name descname"><span class="pre">plotSingleR</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_cluster.plotSingleR" title="Permalink to this definition">¶</a></dt>
<dd><p>Make singleR heatmaps for the references present on the cellhub API.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_cluster.summariseSingleR">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_cluster.</span></span><span class="sig-name descname"><span class="pre">summariseSingleR</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_cluster.summariseSingleR" title="Permalink to this definition">¶</a></dt>
<dd><p>Collect the single R plots into a section for the Summary report.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_cluster.plotGroupNumbers">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_cluster.</span></span><span class="sig-name descname"><span class="pre">plotGroupNumbers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infiles</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_cluster.plotGroupNumbers" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot statistics on cells by group, e.g. numbers of cells per cluster.</p>
<p>Plots are defined on a case-by-case basis in the yaml.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_cluster.clusterStats">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_cluster.</span></span><span class="sig-name descname"><span class="pre">clusterStats</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_cluster.clusterStats" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute per-cluster statistics (e.g. mean expression level).</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_cluster.findMarkers">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_cluster.</span></span><span class="sig-name descname"><span class="pre">findMarkers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_cluster.findMarkers" title="Permalink to this definition">¶</a></dt>
<dd><p>Find per-cluster marker genes. Just execute the rank_genes_groups routine,
no filtering here.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_cluster.summariseMarkers">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_cluster.</span></span><span class="sig-name descname"><span class="pre">summariseMarkers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infiles</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_cluster.summariseMarkers" title="Permalink to this definition">¶</a></dt>
<dd><p>Summarise the differentially expressed marker genes. P-values
are adjusted per-cluster are filtering out genes with low expression
levels and fold changes. Per-cluster gene universes are prepared for
the pathway analysis.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_cluster.topMarkerHeatmap">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_cluster.</span></span><span class="sig-name descname"><span class="pre">topMarkerHeatmap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infiles</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_cluster.topMarkerHeatmap" title="Permalink to this definition">¶</a></dt>
<dd><p>Make the top marker heatmap.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_cluster.dePlots">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_cluster.</span></span><span class="sig-name descname"><span class="pre">dePlots</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_cluster.dePlots" title="Permalink to this definition">¶</a></dt>
<dd><p>Make per-cluster diagnoistic differential expression plots
(MA and volcano plots).</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_cluster.markerPlots">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_cluster.</span></span><span class="sig-name descname"><span class="pre">markerPlots</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infiles</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_cluster.markerPlots" title="Permalink to this definition">¶</a></dt>
<dd><p>Make the per-cluster marker plots
TODO: add some version of split dot plots back..</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_cluster.plotMarkerNumbers">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_cluster.</span></span><span class="sig-name descname"><span class="pre">plotMarkerNumbers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_cluster.plotMarkerNumbers" title="Permalink to this definition">¶</a></dt>
<dd><p>Summarise the numbers of marker genes for each cluster.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_cluster.markers">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_cluster.</span></span><span class="sig-name descname"><span class="pre">markers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_cluster.markers" title="Permalink to this definition">¶</a></dt>
<dd><p>Target to run marker gene plotting tasks.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_cluster.parseGMTs">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_cluster.</span></span><span class="sig-name descname"><span class="pre">parseGMTs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">param_keys</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['gmt_pathway_files_']</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_cluster.parseGMTs" title="Permalink to this definition">¶</a></dt>
<dd><p>Helper function for parsing the lists of GMT files</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_cluster.genesetAnalysis">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_cluster.</span></span><span class="sig-name descname"><span class="pre">genesetAnalysis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_cluster.genesetAnalysis" title="Permalink to this definition">¶</a></dt>
<dd><p>Naive geneset over-enrichment analysis of cluster marker genes.</p>
<p>Testing is performed with the gsfisher package.</p>
<p>GO categories and KEGG pathways are tested by default.</p>
<p>Arbitrary sets of genes cat be supplied as GMT files
(e.g. such as those from MSigDB).</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_cluster.summariseGenesetAnalysis">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_cluster.</span></span><span class="sig-name descname"><span class="pre">summariseGenesetAnalysis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_cluster.summariseGenesetAnalysis" title="Permalink to this definition">¶</a></dt>
<dd><p>Summarise the geneset over-enrichment analyses of cluster marker genes.</p>
<p>Enriched pathways are saved as dotplots and exported in excel format.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_cluster.genesets">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_cluster.</span></span><span class="sig-name descname"><span class="pre">genesets</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_cluster.genesets" title="Permalink to this definition">¶</a></dt>
<dd><p>Intermediate target to run geneset tasks</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_cluster.plots">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_cluster.</span></span><span class="sig-name descname"><span class="pre">plots</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_cluster.plots" title="Permalink to this definition">¶</a></dt>
<dd><p>Target to run all the plotting functions.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_cluster.latexVars">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_cluster.</span></span><span class="sig-name descname"><span class="pre">latexVars</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infiles</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_cluster.latexVars" title="Permalink to this definition">¶</a></dt>
<dd><p>Prepare a file containing the latex variable definitions for the reports.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_cluster.summaryReportSource">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_cluster.</span></span><span class="sig-name descname"><span class="pre">summaryReportSource</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_cluster.summaryReportSource" title="Permalink to this definition">¶</a></dt>
<dd><p>Write the latex source for the summary report.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_cluster.summaryReport">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_cluster.</span></span><span class="sig-name descname"><span class="pre">summaryReport</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_cluster.summaryReport" title="Permalink to this definition">¶</a></dt>
<dd><p>Compile the summary report to PDF format.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_cluster.markerReportSource">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_cluster.</span></span><span class="sig-name descname"><span class="pre">markerReportSource</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_cluster.markerReportSource" title="Permalink to this definition">¶</a></dt>
<dd><p>Write the latex source file  for the marker report.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_cluster.markerReport">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_cluster.</span></span><span class="sig-name descname"><span class="pre">markerReport</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_cluster.markerReport" title="Permalink to this definition">¶</a></dt>
<dd><p>Prepare a PDF report visualising the discovered cluster markers.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_cluster.export">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_cluster.</span></span><span class="sig-name descname"><span class="pre">export</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_cluster.export" title="Permalink to this definition">¶</a></dt>
<dd><p>Link output files to a directory in the “reports.dir” folder.</p>
<p>Prepare folders containing the reports, differentially expressed genes
and geneset tables for each analysis.</p>
<p>TODO: link in the cellxgene anndata files.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_cluster.report">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_cluster.</span></span><span class="sig-name descname"><span class="pre">report</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_cluster.report" title="Permalink to this definition">¶</a></dt>
<dd><p>Target for building the reports.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_cluster.cellxgene">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_cluster.</span></span><span class="sig-name descname"><span class="pre">cellxgene</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_cluster.cellxgene" title="Permalink to this definition">¶</a></dt>
<dd><p>Export an anndata object for cellxgene.</p>
</dd></dl>

</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">cellhub</a></h1>








<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../Installation.html">Installation</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../WorkflowOverview.html">Workflow Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Usage.html">Usage</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../IFNbExample.html">IFNb PBMC example</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../CodingGuidelines.html">Coding Guidelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="../API.html">API</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../tasks/parameters.html">Parameters submodule</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tasks/api.html">API</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tasks/cellbender.html">Cellbender tasks</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="pipeline_adt_norm.html">Pipeline ADT normalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="pipeline_ambient_rna.html">Pipeline ambient rna</a></li>
<li class="toctree-l1"><a class="reference internal" href="pipeline_annotation.html">Pipeline annotation</a></li>
<li class="toctree-l1"><a class="reference internal" href="pipeline_cellbender.html">Pipeline cellbender</a></li>
<li class="toctree-l1"><a class="reference internal" href="pipeline_celldb.html">Pipeline celldb</a></li>
<li class="toctree-l1"><a class="reference internal" href="pipeline_cellranger_multi.html">Pipeline Cellranger Multi</a></li>
<li class="toctree-l1"><a class="reference internal" href="pipeline_cell_qc.html">Pipeline Cell QC</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Pipeline cluster</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#overview">Overview</a></li>
<li class="toctree-l2"><a class="reference internal" href="#usage">Usage</a></li>
<li class="toctree-l2"><a class="reference internal" href="#pipeline-output">Pipeline output</a></li>
<li class="toctree-l2"><a class="reference internal" href="#code">Code</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="pipeline_dehash.html">Pipeline dehash</a></li>
<li class="toctree-l1"><a class="reference internal" href="pipeline_emptydrops.html">Pipeline Emptydrops</a></li>
<li class="toctree-l1"><a class="reference internal" href="pipeline_fetch_cells.html">Pipeline fetch cells</a></li>
<li class="toctree-l1"><a class="reference internal" href="pipeline_singleR.html">Pipeline singleR</a></li>
<li class="toctree-l1"><a class="reference internal" href="pipeline_velocyto.html">Pipeline Velocyto</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
      <li>Previous: <a href="pipeline_cell_qc.html" title="previous chapter">Pipeline Cell QC</a></li>
      <li>Next: <a href="pipeline_dehash.html" title="next chapter">Pipeline dehash</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2021, Sansom lab.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 5.1.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="../_sources/pipelines/pipeline_cluster.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>