
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Pipeline ADT normalization &#8212; cellhub 0.1 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/alabaster.css" />
    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Pipeline ambient rna" href="pipeline_ambient_rna.html" />
    <link rel="prev" title="API" href="../tasks/api.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <span class="target" id="module-cellhub.pipeline_adt_norm"></span><section id="pipeline-adt-normalization">
<h1>Pipeline ADT normalization<a class="headerlink" href="#pipeline-adt-normalization" title="Permalink to this heading">¶</a></h1>
<section id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this heading">¶</a></h2>
<p>This pipeline implements three normalization methods:</p>
<ul class="simple">
<li><p>DSB (<a class="reference external" href="https://www.biorxiv.org/content/10.1101/2020.02.24.963603v3">https://www.biorxiv.org/content/10.1101/2020.02.24.963603v3</a>)</p></li>
<li><p>Median-based (<a class="reference external" href="https://bioconductor.org/books/release/OSCA/integrating-with-protein-abundance.html">https://bioconductor.org/books/release/OSCA/integrating-with-protein-abundance.html</a>)</p></li>
<li><p>CLR (<a class="reference external" href="https://satijalab.org/seurat/archive/v3.0/multimodal_vignette.html">https://satijalab.org/seurat/archive/v3.0/multimodal_vignette.html</a>)</p></li>
</ul>
<section id="configuration">
<h3>Configuration<a class="headerlink" href="#configuration" title="Permalink to this heading">¶</a></h3>
<p>The pipeline requires a configured <code class="file docutils literal notranslate"><span class="pre">pipeline_adt_norm.yml</span></code> file. Default configuration files can be generated by executing:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="o">&lt;</span><span class="n">srcdir</span><span class="o">&gt;/</span><span class="n">pipeline_adt_norm</span><span class="o">.</span><span class="n">py</span> <span class="n">config</span>
</pre></div>
</div>
</section>
<section id="input-files">
<h3>Input files<a class="headerlink" href="#input-files" title="Permalink to this heading">¶</a></h3>
<p>This pipeline requires the unfiltered gene-expression and ADT count matrices 
and a list of high quality barcodes most likely representing single-cells.</p>
<p>This means that ideally this pipeline is run after high quality cells are selected
via the pipeline_fetch_cells.py.</p>
<p>This pipeline will look for the unfiltered matrix in the api:</p>
<blockquote>
<div><p>./api/cellranger.multi/ADT/unfiltered/<em>/mtx/</em>.gz</p>
<p>./api/cellranger.multi/GEX/unfiltered/<em>/mtx/</em>.gz</p>
</div></blockquote>
</section>
<section id="dependencies">
<h3>Dependencies<a class="headerlink" href="#dependencies" title="Permalink to this heading">¶</a></h3>
<p>This pipeline requires:
* cgat-core: <a class="reference external" href="https://github.com/cgat-developers/cgat-core">https://github.com/cgat-developers/cgat-core</a>
* R dependencies required in the r scripts</p>
</section>
</section>
<section id="pipeline-output">
<h2>Pipeline output<a class="headerlink" href="#pipeline-output" title="Permalink to this heading">¶</a></h2>
<p>The pipeline returns a adt_norm.dir folder containing one folder per methodology
adt_dsb.dir, adt_median.dir, and adt_clr.dir with per-sample folders 
conatining market matrices [features, qc-barcodes] with the normalized values.</p>
</section>
<section id="code">
<h2>Code<a class="headerlink" href="#code" title="Permalink to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_adt_norm.gexdepth">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_adt_norm.</span></span><span class="sig-name descname"><span class="pre">gexdepth</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_adt_norm.gexdepth" title="Permalink to this definition">¶</a></dt>
<dd><p>This task will run R/adt_calculate_depth_dist.R,
It will describe the GEX UMI distribution of the background and cell-containing 
barcodes. This will help to assess the quality of the ADT data and will inform
about the definition of the background barcodes.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_adt_norm.gexdepthAPI">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_adt_norm.</span></span><span class="sig-name descname"><span class="pre">gexdepthAPI</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infiles</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_adt_norm.gexdepthAPI" title="Permalink to this definition">¶</a></dt>
<dd><p>Add the umi depth metrics results to the API</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_adt_norm.adtdepth">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_adt_norm.</span></span><span class="sig-name descname"><span class="pre">adtdepth</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_adt_norm.adtdepth" title="Permalink to this definition">¶</a></dt>
<dd><p>This task will run R/adt_calculate_depth_dist.R,
It will describe the ADT UMI distribution of the background and cell-containing 
barcodes. This will help to assess the quality of the ADT data and will inform
the definition of the background barcodes.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_adt_norm.adtdepthAPI">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_adt_norm.</span></span><span class="sig-name descname"><span class="pre">adtdepthAPI</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infiles</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_adt_norm.adtdepthAPI" title="Permalink to this definition">¶</a></dt>
<dd><p>Add the umi depth metrics results to the API</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_adt_norm.adt_plot_norm">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_adt_norm.</span></span><span class="sig-name descname"><span class="pre">adt_plot_norm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_adt_norm.adt_plot_norm" title="Permalink to this definition">¶</a></dt>
<dd><p>This task will run R/adt_plot_norm.R,
It will create a visual report on the cell vs background dataset split and,
if the user provided GEX and ADT UMI thresholds, those will be included.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_adt_norm.dsb_norm">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_adt_norm.</span></span><span class="sig-name descname"><span class="pre">dsb_norm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_adt_norm.dsb_norm" title="Permalink to this definition">¶</a></dt>
<dd><p>This task runs R/adt_normalize.R.
It reads the unfiltered ADT count matrix and calculates DSB normalized ADT 
expression matrix which is then saved like market matrices per sample.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_adt_norm.dsbAPI">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_adt_norm.</span></span><span class="sig-name descname"><span class="pre">dsbAPI</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_adt_norm.dsbAPI" title="Permalink to this definition">¶</a></dt>
<dd><p>Register the ADT normalized mtx files on the API endpoint</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_adt_norm.median_norm">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_adt_norm.</span></span><span class="sig-name descname"><span class="pre">median_norm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_adt_norm.median_norm" title="Permalink to this definition">¶</a></dt>
<dd><p>This task runs R/adt_get_median_normalization.R,
It reads the filtered ADT count matrix and performed median-based 
normalization. Calculates median-based normalized ADT expression matrix and
writes market matrices per sample.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_adt_norm.medianAPI">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_adt_norm.</span></span><span class="sig-name descname"><span class="pre">medianAPI</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_adt_norm.medianAPI" title="Permalink to this definition">¶</a></dt>
<dd><p>Register the ADT normalized mtx files on the API endpoint</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_adt_norm.clr_norm">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_adt_norm.</span></span><span class="sig-name descname"><span class="pre">clr_norm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_adt_norm.clr_norm" title="Permalink to this definition">¶</a></dt>
<dd><p>This task runs R/get_median_clr_normalization.R,
It reads the filtered ADT count matrix and performes CLR 
normalization. Writes market matrices per sample.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_adt_norm.clrAPI">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_adt_norm.</span></span><span class="sig-name descname"><span class="pre">clrAPI</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_adt_norm.clrAPI" title="Permalink to this definition">¶</a></dt>
<dd><p>Register the CLR-normalized ADT mtx files on the API endpoint</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_adt_norm.plot">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_adt_norm.</span></span><span class="sig-name descname"><span class="pre">plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">infile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_adt_norm.plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Draw the pipeline flowchart</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cellhub.pipeline_adt_norm.full">
<span class="sig-prename descclassname"><span class="pre">cellhub.pipeline_adt_norm.</span></span><span class="sig-name descname"><span class="pre">full</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cellhub.pipeline_adt_norm.full" title="Permalink to this definition">¶</a></dt>
<dd><p>Run the full pipeline.</p>
</dd></dl>

</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">cellhub</a></h1>








<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../Installation.html">Installation</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../WorkflowOverview.html">Workflow Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Usage.html">Usage</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../IFNbExample.html">IFNb PBMC example</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../CodingGuidelines.html">Coding Guidelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="../API.html">API</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../tasks/api.html">API</a></li>
</ul>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Pipeline ADT normalization</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#overview">Overview</a></li>
<li class="toctree-l2"><a class="reference internal" href="#pipeline-output">Pipeline output</a></li>
<li class="toctree-l2"><a class="reference internal" href="#code">Code</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="pipeline_ambient_rna.html">Pipeline ambient rna</a></li>
<li class="toctree-l1"><a class="reference internal" href="pipeline_annotation.html">Pipeline annotation</a></li>
<li class="toctree-l1"><a class="reference internal" href="pipeline_cellbender.html">Pipeline cellbender</a></li>
<li class="toctree-l1"><a class="reference internal" href="pipeline_celldb.html">Pipeline celldb</a></li>
<li class="toctree-l1"><a class="reference internal" href="pipeline_cellranger_multi.html">Pipeline Cellranger Multi</a></li>
<li class="toctree-l1"><a class="reference internal" href="pipeline_cell_qc.html">Pipeline Cell QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="pipeline_cluster.html">Pipeline cluster</a></li>
<li class="toctree-l1"><a class="reference internal" href="pipeline_dehash.html">Pipeline dehash</a></li>
<li class="toctree-l1"><a class="reference internal" href="pipeline_emptydrops.html">Pipeline Emptydrops</a></li>
<li class="toctree-l1"><a class="reference internal" href="pipeline_fetch_cells.html">Pipeline fetch cells</a></li>
<li class="toctree-l1"><a class="reference internal" href="pipeline_singleR.html">Pipeline singleR</a></li>
<li class="toctree-l1"><a class="reference internal" href="pipeline_velocyto.html">Pipeline Velocyto</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
      <li>Previous: <a href="../tasks/api.html" title="previous chapter">API</a></li>
      <li>Next: <a href="pipeline_ambient_rna.html" title="next chapter">Pipeline ambient rna</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2021, Sansom lab.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 5.1.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="../_sources/pipelines/pipeline_adt_norm.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>